File: seafood_delivery_app/backend/services/route_optimizer.py
"""
Nearest-Neighbor route optimizer.

Given a driver's current GPS position and a list of delivery stops,
returns the stops ordered so the driver always visits the nearest
unvisited stop next.  This is a greedy O(n²) TSP approximation that
works well for the small batches (≤ 10 orders) typical in delivery.
"""
import math
from typing import List, Dict, Any


def haversine_meters(lat1: float, lng1: float, lat2: float, lng2: float) -> float:
    """Return the great-circle distance in metres between two GPS coordinates."""
    R = 6_371_000  # Earth radius in metres
    phi1, phi2 = math.radians(lat1), math.radians(lat2)
    d_phi = math.radians(lat2 - lat1)
    d_lam = math.radians(lng2 - lng1)
    a = math.sin(d_phi / 2) ** 2 + math.cos(phi1) * math.cos(phi2) * math.sin(d_lam / 2) ** 2
    return 2 * R * math.atan2(math.sqrt(a), math.sqrt(1 - a))


def optimize_route(
    driver_lat: float,
    driver_lng: float,
    stops: List[Dict[str, Any]],
) -> List[Dict[str, Any]]:
    """
    Parameters
    ----------
    driver_lat / driver_lng : current driver GPS position
    stops : list of dicts, each must contain at minimum:
            { "order_id": int, "delivery_lat": float, "delivery_lng": float }
            Any extra keys are passed through unchanged.

    Returns
    -------
    A new list in optimised visit order.  Each dict gains two extra keys:
      "sequence"              : 1-based position in the route
      "distance_from_prev_m"  : Haversine metres from the previous stop
                                (or from driver position for the first stop)
    """
    if not stops:
        return []

    remaining = list(stops)
    route: List[Dict[str, Any]] = []
    cur_lat, cur_lng = driver_lat, driver_lng

    while remaining:
        nearest = min(
            remaining,
            key=lambda s: haversine_meters(cur_lat, cur_lng, s["delivery_lat"], s["delivery_lng"]),
        )
        dist = haversine_meters(cur_lat, cur_lng, nearest["delivery_lat"], nearest["delivery_lng"])
        stop = dict(nearest)
        stop["sequence"] = len(route) + 1
        stop["distance_from_prev_m"] = round(dist)
        route.append(stop)
        remaining.remove(nearest)
        cur_lat, cur_lng = nearest["delivery_lat"], nearest["delivery_lng"]

    return route

File: seafood_delivery_app/driver_app/lib/services/route_optimizer.dart
import 'dart:math';
import 'package:latlong2/latlong.dart';
import '../models/assigned_order_model.dart';
import '../models/route_stop_model.dart';

/// Client-side Nearest-Neighbor route optimizer.
///
/// Given the driver's current GPS position and a list of delivery orders,
/// returns a new list sorted so the driver always moves to the
/// closest unvisited stop first.  Pure Dart — no network call needed.
class RouteOptimizer {
  /// Returns the optimised list of [RouteStop]s for the given orders.
  static List<RouteStop> optimize({
    required LatLng driverPosition,
    required List<AssignedOrder> orders,
  }) {
    if (orders.isEmpty) return [];

    final unvisited = orders.where((o) => o.hasCoordinates).toList();
    final result = <RouteStop>[];
    LatLng current = driverPosition;

    while (unvisited.isNotEmpty) {
      AssignedOrder nearest = unvisited.reduce(
        (a, b) => _dist(current, LatLng(a.deliveryLat!, a.deliveryLng!)) <=
                _dist(current, LatLng(b.deliveryLat!, b.deliveryLng!))
            ? a
            : b,
      );
      final dest = LatLng(nearest.deliveryLat!, nearest.deliveryLng!);
      final dist = _dist(current, dest);
      result.add(RouteStop(
        orderId: nearest.orderId,
        sequence: result.length + 1,
        deliveryAddress: nearest.deliveryAddress,
        location: dest,
        distanceFromPrevMeters: dist,
        status: nearest.status,
        total: nearest.total,
      ));
      unvisited.remove(nearest);
      current = dest;
    }
    return result;
  }

  /// Re-runs after a delivery is completed — removes that stop and re-sequences.
  static List<RouteStop> reOptimize({
    required LatLng driverPosition,
    required List<AssignedOrder> remainingOrders,
  }) =>
      optimize(driverPosition: driverPosition, orders: remainingOrders);

  /// Haversine great-circle distance in metres.
  static double _dist(LatLng a, LatLng b) {
    const R = 6371000.0;
    final dLat = _rad(b.latitude - a.latitude);
    final dLng = _rad(b.longitude - a.longitude);
    final lat1 = _rad(a.latitude);
    final lat2 = _rad(b.latitude);
    final x = sin(dLat / 2) * sin(dLat / 2) +
        cos(lat1) * cos(lat2) * sin(dLng / 2) * sin(dLng / 2);
    return 2 * R * atan2(sqrt(x), sqrt(1 - x));
  }

  static double _rad(double deg) => deg * pi / 180;

  /// Returns distance in metres between driver and a specific order's address.
  static double distanceTo(LatLng driver, AssignedOrder order) {
    if (!order.hasCoordinates) return double.infinity;
    return _dist(driver, LatLng(order.deliveryLat!, order.deliveryLng!));
  }

  /// Returns true if driver is within [thresholdMeters] of [destination].
  static bool isWithinProximity(LatLng driver, LatLng destination, double thresholdMeters) =>
      _dist(driver, destination) <= thresholdMeters;
}
